adapters/triplelift_native/triplelift_native.go

```diff
@@ -0,0 +1,195 @@
+	if tlext.Floor == nil {
+		return nil
+	} else {
+		imp.BidFloor = *tlext.Floor
+	}
+	// no error
+	return nil
+}
+
+// Returns the effective publisher ID
+func effectivePubID(pub *openrtb.Publisher) string {
+	if pub != nil {
+		if pub.Ext != nil {
+			var pubExt openrtb_ext.ExtPublisher
+			err := json.Unmarshal(pub.Ext, &pubExt)
+			if err == nil && pubExt.Prebid != nil && pubExt.Prebid.ParentAccount != nil && *pubExt.Prebid.ParentAccount != "" {
+				return *pubExt.Prebid.ParentAccount
+			}
+		}
+		if pub.ID != "" {
+			return pub.ID
+		}
+	}
+	return "unknown"
+}
+
+func (a *TripleliftNativeAdapter) MakeRequests(request *openrtb.BidRequest, extra *adapters.ExtraRequestInfo) ([]*adapters.RequestData, []error) {
+	errs := make([]error, 0, len(request.Imp)+1)
+	reqs := make([]*adapters.RequestData, 0, 1)
+	// copy the request, because we are going to mutate it
+	tlRequest := *request
+	// this will contain all the valid impressions
+	var validImps []openrtb.Imp
+	// pre-process the imps
+	for _, imp := range tlRequest.Imp {
+		if err := processImp(&imp); err == nil {
+			validImps = append(validImps, imp)
+		} else {
+			errs = append(errs, err)
+		}
+	}
+	publisher := getPublisher(request)
+	publisherID := effectivePubID(publisher)
+	if _, exists := a.extInfo.PublisherWhitelistMap[publisherID]; !exists {
+		err := fmt.Errorf("Unsupported publisher for triplelift_native")
+		return nil, []error{err}
+	}
+	if len(validImps) == 0 {
+		err := fmt.Errorf("No valid impressions for triplelift")
+		errs = append(errs, err)
+		return nil, errs
+	}
+	tlRequest.Imp = validImps
+	reqJSON, err := json.Marshal(tlRequest)
+	if err != nil {
+		errs = append(errs, err)
+		return nil, errs
+	}
+	headers := http.Header{}
+	headers.Add("Content-Type", "application/json;charset=utf-8")
+	headers.Add("Accept", "application/json")
+	ad := a.endpoint
+	reqs = append(reqs, &adapters.RequestData{
+		Method:  "POST",
+		Uri:     ad,
+		Body:    reqJSON,
+		Headers: headers})
+	return reqs, errs
+}
+
+func getPublisher(request *openrtb.BidRequest) *openrtb.Publisher {
+	if request.App != nil {
+		return request.App.Publisher
+	}
+	return request.Site.Publisher
+}
+
+func getBidCount(bidResponse openrtb.BidResponse) int {
+	c := 0
+	for _, sb := range bidResponse.SeatBid {
+		c = c + len(sb.Bid)
+	}
+	return c
+}
+
+func (a *TripleliftNativeAdapter) MakeBids(internalRequest *openrtb.BidRequest, externalRequest *adapters.RequestData, response *adapters.ResponseData) (*adapters.BidderResponse, []error) {
+	if response.StatusCode == http.StatusNoContent {
+		return nil, nil
+	}
+
+	if response.StatusCode == http.StatusBadRequest {
+		return nil, []error{&errortypes.BadInput{
+			Message: fmt.Sprintf("Unexpected status code: %d. Run with request.debug = 1 for more info", response.StatusCode),
+		}}
+	}
+
+	if response.StatusCode != http.StatusOK {
+		return nil, []error{fmt.Errorf("Unexpected status code: %d. Run with request.debug = 1 for more info", response.StatusCode)}
```

Comment:
Would be good to change this error to be of type `errortypes.BadServerResponse`. Specific error types help better understand what went wrong and also helps in taking better actions based on error codes for these specific error types. Specially, when the API is being called by another system/service.

Reply:
makes sense; done